utils/validators.ts

```diff
@@ -0,0 +1,259 @@
+import { ObjectId } from 'mongodb'
+
+interface Property {
+  key: string
+  types: string
+  required: boolean
+}
+
+export interface ValidationResult {
+  success: boolean
+  message?: string
+}
+
+/**
+ * Checks the validity of an Attribute object. Does not validate child objects.
+ * @param attribute The Attribute object to test
+ * @returns A ValidationResult object
+ */
+export function validateAttribute(attribute: any) {
+  return validateProperties(attributeModelProperties, attribute)
+}
+
+/**
+ * Checks the validity of an Category object. Does not validate child objects.
+ * @param category The Category object to test
+ * @returns A ValidationResult object
+ */
+export function validateCategory(category: any) {
+  return validateProperties(categoryModelProperties, category)
+}
+
+/**
+ * Checks the validity of an ItemDefinition object. Does not validate child objects.
+ * @param itemDefinition The ItemDefinition object to test
+ * @returns A ValidationResult object
+ */
+export function validateItemDefinition(itemDefinition: any) {
+  return validateProperties(itemDefinitionModelProperties, itemDefinition)
+}
+
+/**
+ * Checks the validity of an InventoryItem object. Does not validate child objects.
+ * @param inventoryItem The InventoryItem object to test
+ * @returns A ValidationResult object
+ */
+export function validateInventoryItem(inventoryItem: any) {
+  return validateProperties(inventoryItemModelProperties, inventoryItem)
+}
+
+/**
+ * Checks the validity of an User object. Does not validate child objects.
+ * @param user The User object to test
+ * @returns A ValidationResult object
+ */
+export function validateUser(user: any) {
+  return validateProperties(userModelProperties, user)
+}
+
+/**
+ * Validates a given id to ensure it is a valid ObjectId. Throws an error if invalid.
+ * @param id - The ObjectId to validate
+ * @returns true if the objectId was valid, false if not
+ */
+export function validateObjectId(id: string) {
+  if (!id) return false
+  if (ObjectId.isValid(id)) {
+    if (String(new ObjectId(id)) === id) return true
+  }
+  return false
+}
+
+/**
+ * Generic object validator that checks for required fields, extra fields, and type mismatches
+ * @param modelProperties An array of Property objects
+ * @param obj The object to validate
+ * @returns A ValidationResult object
+ */
+function validateProperties(
+  modelProperties: Property[],
+  obj: any
+): ValidationResult {
+  let result: ValidationResult = {
+    success: false,
+  }
+
+  if (typeof obj !== 'object') {
+    result.message = 'Validation target is not an instance of an object'
+    return result
+  }
+
+  // ensures all required properties are present
+  let isValidObject = true
+  for (const modelProp of modelProperties) {
+    if (
+      isValidObject &&
+      modelProp.required &&
+      !obj.hasOwnProperty(modelProp.key)
+    ) {
+      result.message = `Missing required property '${modelProp.key}'`
```

Comment:
is there a way to list that multiple properties are missing? when I try to create an `ItemDefinition` without a `lowStockThreshold` and without a `criticalStockThreshold` the message returned is `{"success":false,"message":"Bad Request Body: Missing required property 'lowStockThreshold'"}`

Reply:
added